### PR TITLE
Separate permit logic and fee

### PR DESCRIPTION
**Pull request by hihiben**
- [x] Extract permit logics
- [x] Disallow permit from original logics
- [x] Implement permit fee
- [x] Update unit tests
- [x] Update typed data tests

### CODE REVIEW

File: src/AgentImplementation.sol [hunk 1]
```diff
@@ -280,4 +351,8 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
             }
         }
     }
+
+    function _calculateFee(uint256 feeRate, uint256 amount) internal pure returns (uint256) {
+        return (amount * feeRate) / (_BPS_BASE + feeRate);
```
Comment (Jeff-CCH) [inline src/AgentImplementation.sol:356]: Wondering why we need to add `feeRate` in `_BPS_BASE + feeRate`?

File: src/AgentImplementation.sol [hunk 2]
```diff
@@ -249,6 +317,9 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
             address token = fees[i].token;
             if (token == _NATIVE) {
                 payable(feeCollector).sendValue(amount);
+            } else if (chargeFromUser) {
+                address user = IRouter(router).currentUser();
+                IAllowanceTransfer(permit2).transferFrom(user, feeCollector, uint160(amount), token);
```
Comment (Jeff-CCH) [inline src/AgentImplementation.sol:322]: Do we need to use `SafeCast` for `uint160(amount)`?

File: src/AgentImplementation.sol [hunk 3]
```diff
@@ -232,7 +300,7 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
         // TODO: implement new msg.value fee charge mechanism
     }
 
-    function _chargeFee(IParam.Fee[] memory fees) internal {
+    function _chargeFee(IParam.Fee[] memory fees, bool chargeFromUser) internal {
         uint256 length = fees.length;
         if (length == 0) return;
```
Comment (Jeff-CCH) [inline src/AgentImplementation.sol:305]: Though this is not modified in this PR, want to be sure if we need `uint256 length = fees.length`? (Seems `fees.length` is already in memory)

File: src/AgentImplementation.sol [hunk 4]
```diff
@@ -59,37 +66,95 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
...
+            } else if (sig == 0x0d58b1db) {
+                // transferFrom((address,address,uint160,address)[])
+                permit2.functionCall(permitLogic, 'ERROR_PERMIT2_TF');
+                if (shouldChargeFee) {
+                    uint256 feeRate = IRouter(router).feeRate();
+                    IAllowanceTransfer.AllowanceTransferDetails[] memory details = abi.decode(
+                        permitLogic[4:],
+                        (IAllowanceTransfer.AllowanceTransferDetails[])
+                    );
+                    uint256 detailsLength = details.length;
```
Comment (Jeff-CCH) [inline src/AgentImplementation.sol:129]: same as the length discussion in `_chargeFee`

File: src/AgentImplementation.sol [hunk 5]
```diff
@@ -59,37 +66,95 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
...
+                        unchecked {
+                            ++j;
+                        }
+                    }
+                    _chargeFee(fees, false);
+                }
+            } else if (sig == 0x2b67b570 || sig == 0x2a2d80d1) {
+                permit2.functionCall(permitLogic, 'ERROR_PERMIT2_P');
+                // permit(address,((address,uint160,uint48,uint48),address,uint256),bytes)
+                // permit(address,((address,uint160,uint48,uint48)[],address,uint256),bytes)
```
Comment (Jeff-CCH) [inline src/AgentImplementation.sol:147]: could move the comments on top of L145 to be consistent with L108 & L121

File: src/Router.sol [hunk 6]
```diff
@@ -442,4 +459,11 @@ contract Router is IRouter, Ownable, EIP712 {
         pauser = pauser_;
         emit PauserSet(pauser_);
     }
+
+    /// @notice Set a new fee rate by owner
+    /// @param feeRate_ The new fee rate in basis points
+    function setFeeRate(uint256 feeRate_) public onlyOwner {
```
[the remark this answers is not in the document]
Comment (hihiben) [reply src/Router.sol:465]: Yes we should, but I'm too lazy to fix all the tests at this moment lol. Will do.

File: src/AgentImplementation.sol (hunk 1 above)
Comment (hihiben) [reply src/AgentImplementation.sol:356]: hmm good question. Maybe it's needed in the previous implementation but not now. Let me think about that.

File: src/AgentImplementation.sol (hunk 3 above)
Comment (hihiben) [reply src/AgentImplementation.sol:305]: Will check.

Comment (hihiben) [reply src/AgentImplementation.sol:305]: https://sandbox.tenderly.co/hihiben/curved-monitor
Seems to be slightly cheaper.

File: test/integration/ERC1155Market.t.sol [hunk 7]
```diff
@@ -63,16 +63,19 @@ contract ERC1155MarketTest is Test, ERC20Permit2Utils, ERC1155Utils {
         uint256 amountIn = amount * market.amount();
         deal(address(tokenIn), user, amountIn);
 
+        // Encode permit2Datas
+        bytes[] memory datas = new bytes[](1);
```
Comment (Jeff-CCH) [inline test/integration/ERC1155Market.t.sol:67]: data is already a plural. `s` is not needed

File: test/Agent.t.sol [hunk 8]
```diff
@@ -78,6 +79,47 @@ contract AgentTest is Test {
...
+        // Mock the call to permit2
+        vm.etch(permit2, address(agent).code);
+        vm.mockCall(permit2, 0, datas[0], '');
+        // Mock the call to router getting fee rate and fee collector
+        vm.etch(router, address(agent).code);
+        vm.mockCall(router, 0, abi.encodeCall(IRouter.feeRate, ()), abi.encode(0));
+        vm.mockCall(router, 0, abi.encodeCall(IRouter.feeCollector, ()), abi.encode(makeAddr('collector')));
+        agent.execute(datas, logicsEmpty, tokensReturnEmpty);
+        agent.executeWithSignerFee(datas, logicsEmpty, feesEmpty, tokensReturnEmpty);
+    }
```
Comment (Jeff-CCH) [inline test/Agent.t.sol:121]: Please help add `vm.stopPrank();` at the end

File: test/integration/Permit2FeeCharging.t.sol [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    IParam.Logic[] public logicsEmpty;
+
+    function setUp() external {
+        (user, userPrivateKey) = makeAddrAndKey('User');
+        (signer, signerPrivateKey) = makeAddrAndKey('Signer');
+        feeCollector = makeAddr('FeeCollector');
+        address pauser = makeAddr('Pauser');
+
+        // Depoly contracts
+        router = new Router(makeAddr('WrappedNative'), PERMIT2_ADDR, address(this), pauser, feeCollector);
```
Comment (Jeff-CCH) [inline test/integration/Permit2FeeCharging.t.sol:42]: `pauser` is not used in the test so we can use `makeAddr('Pauser')` directly in `new Router()`.

File: test/integration/Permit2FeeCharging.t.sol [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        router.addSigner(signer);
+
+        // Setup permit2
+        erc20Permit2UtilsSetUp(user, userPrivateKey, address(userAgent));
+        permitToken(IERC20(USDC));
+
+        vm.label(address(router), 'Router');
+        vm.label(address(userAgent), 'UserAgent');
+        vm.label(feeCollector, 'FeeCollector');
+        vm.label(PERMIT2_ADDR, 'Permit2Address');
```
Comment (Jeff-CCH) [inline test/integration/Permit2FeeCharging.t.sol:54]: `Permit2Address` -> `Permit2`
I think Permit2 is clear enough.

File: test/integration/Permit2FeeCharging.t.sol [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        amount = bound(amount, 0, (IERC20(USDC).totalSupply() * (BPS_BASE - feeRate)) / BPS_BASE);
+
+        // Set fee rate
+        router.setFeeRate(feeRate);
+
+        // Encode logic
+        bytes[] memory datas = new bytes[](1);
+
+        uint256 amountWithFee = (amount * (BPS_BASE + feeRate)) / BPS_BASE;
+
```
Comment (Jeff-CCH) [inline test/integration/Permit2FeeCharging.t.sol:68]: Do we need the empty line for L66 & L68?

File: src/Router.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (hihiben) [reply src/Router.sol:465]: Decided to so an extra step if setting fee rate is necessary for the deploying chain. Will not change the constructor at this moment.

File: test/integration/Permit2FeeCharging.t.sol (hunk 10 above)
Comment (hihiben) [reply test/integration/Permit2FeeCharging.t.sol]: Changed.

File: test/integration/Permit2FeeCharging.t.sol (hunk 9 above)
Comment (hihiben) [reply test/integration/Permit2FeeCharging.t.sol]: Changed.

File: test/integration/BalancerV2.t.sol [hunk 12]
```diff
@@ -21,6 +21,7 @@ contract BalancerV2IntegrationTest is Test {
     uint256 public constant SIGNER_REFERRAL = 1;
     IBalancerV2Vault public constant balancerV2Vault = IBalancerV2Vault(0xBA12222222228d8Ba445958a75a0704d566BF2C8);
     IERC20 public constant USDC = IERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48);
+    address internal constant permit2Addr = address(0x000000000022D473030F116dDEE9F6B43aC78BA3);
```
[the remark this answers is not in the document]
Comment (hihiben) [reply test/integration/BalancerV2.t.sol]: Removed.

File: test/integration/AaveV3.t.sol [hunk 13]
```diff
@@ -56,6 +56,7 @@ contract AaveV3IntegrationTest is Test {
     IERC20 public constant USDC = IERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48);
     address public constant AUSDC_V3 = 0x98C23E9d8f34FEFb1B7BD6a91B7FF122F4e16F5c;
     IDebtToken public constant AUSDC_V3_DEBT_VARIABLE = IDebtToken(0x72E95b8931767C79bA4EeE721354d6E99a61D004);
+    address internal constant permit2Addr = address(0x000000000022D473030F116dDEE9F6B43aC78BA3);
```
[the remark this answers is not in the document]
Comment (Jeff-CCH) [reply test/integration/AaveV3.t.sol:59]: seems not yet removed lol

File: src/Router.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jeff-CCH) [reply src/Router.sol:465]: I see. Do we need to change the scope from `public` to `external`?

File: src/AgentImplementation.sol [hunk 14]
```diff
@@ -59,37 +68,95 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
...
         _returnTokens(tokensReturn);
     }
 
+    function _doPermit(bytes[] calldata permit2Datas, bool shouldChargeFee) internal {
+        for (uint256 i; i < permit2Datas.length; ) {
+            bytes calldata permit2Data = permit2Datas[i];
+            bytes4 selector = bytes4(permit2Data[:4]);
+            if (selector == 0x36c78516) {
+                // transferFrom(address,address,uint160,address)
+                permit2.functionCall(permit2Data, 'ERROR_PERMIT2_TF');
```
[the remark this answers is not in the document]
Comment (hihiben) [reply src/AgentImplementation.sol:111]: TransferFrom and Permit. Inspired by Uniswap error message.

File: src/AgentImplementation.sol [hunk 15]
```diff
@@ -59,37 +68,95 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
     /// @notice Execute arbitrary logics and is only callable by the router. Charge fee during the execution of
     ///         msg.value, permit2 and flash loans.
     /// @dev The router is designed to prevent reentrancy so additional prevention is not needed here
+    /// @param permit2Datas The logics to be processed through permit2 contract
```
[the remark this answers is not in the document]
Comment (zodahu) [reply src/AgentImplementation.sol]: 1. Please maintain consistent comments for parameter natspec by "Array of datas", or changes all 
```
/// @param logics Array of logics...
/// @param fees Array of fees...
/// @param tokensReturn Array of ERC-20 tokens...
```
2. Here are 4 instances in the `Router` that haven't been modified.
image

File: test/integration/Permit2FeeCharging.t.sol [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    function testChargePermit2TransferFromFeeFromUser(uint256 amount, uint256 feeRate) external {
+        feeRate = bound(feeRate, 0, BPS_BASE - 1);
+        amount = bound(amount, 0, (IERC20(USDC).totalSupply() * (BPS_BASE - feeRate)) / BPS_BASE);
+
+        // Set fee rate
+        router.setFeeRate(feeRate);
+
+        // Encode logic
```
[the remark this answers is not in the document]
Comment (zodahu) [reply test/integration/Permit2FeeCharging.t.sol:100]: This hasn't been fixed.

File: src/AgentImplementation.sol (hunk 15 above)
[the remark this answers is not in the document]
Comment (hihiben) [reply src/AgentImplementation.sol]: Fine i'll use Array of datas. We can correct the grammar afterwards if we want to.

**Review by zodahu — APPROVED** [review]
LGTM

File: src/Router.sol [hunk 17]
```diff
@@ -312,7 +312,7 @@ contract Router is IRouter, Ownable, EIP712 {
             if (block.timestamp > deadline) revert SignatureExpired(deadline);
             if (!user.isValidSignatureNow(_hashTypedDataV4(details._hash()), userSignature)) revert InvalidSignature();
             if (executionNonces[user] != nonce) revert InvalidNonce();
-            executionNonces[user] = nonce + 1;
+            ++executionNonces[user];
```
Comment (Jeff-CCH) [inline src/Router.sol:315]: one more nonce in `src/library/Delegation.sol`. please help fix it too. Thanks.

Comment (hihiben) [reply src/Router.sol:315]: Nonce does not guarantee to be the same in that function. I don't think it's proper to do that.

**Review by Jeff-CCH — APPROVED** [review]
LGTM

**Pull request merged**